accepted/2021/User-Experience/Apple-Silicon-UX.md

```diff
@@ -0,0 +1,254 @@
+
+The experiment:
+
+- Assumed .NET runtime versions less than 6.0 will always require `x86_64` architecture. All other runtime versions will require `arm64` architecture on Apple Silicon.
+
+- Made the .NET 6 `dotnet` & `libhostfxr.dylib` (`hostfxr`) universal binaries. This means both native and emulated processes can perform runtime resolution to identify the correct runtime to use.
+
+- The `hostfxr` was modified to enforce the architecture version  policy above. It will return an architecture mismatch error when the runtime (by version policy above) require a different architecture than the currently running process.
+
+- `dotnet` was modified to enable switching the architecture of the current process when it saw an architecture mismatch error returned by `hostfxr`. The current process will switch architecture as needed and run with the same arguments as the new process. The switching will use `posix_spawn` with changes to `posix_spawnattr_t` to set `POSIX_SPAWN_SETEXEC` and preferred architecture via `posix_spawnattr_setbinpref_np` (see Apple Silicon man page) and setting `sysctl` `kern.curproc_arch_affinity` property appropriately.
+
+Overall the experiment demonstrated a highly workable user experience.
+
+- The experiment represents a viable design alternative. It did not support side-by-side install of multiple architectures of the same runtime version.
+
+- The implementation was incomplete. It did not automate the build process or deal with installer issues
+
+### Dotnet Publish Experience Inconsistent
+
+#### Signing
+
+- dotnet publish for Xamarin workloads handles signing automatically.
+- dotnet publish for CoreClr console apps do not handle signing automatically. The developer would need to explicitly sign after publishing for Apple Silicon.  The signing command is relatively simple `codesign -s <signature> --entitlements <app-entitlements-path> <app-path>`.
+- There is no message to the customer that the published app needs to be signed
+
+#### Dotnet Publish as Universal Binaries
+
+- This scenario is also supported for Xamarin iOS workloads. This is a slightly different scenario as Xamarin iOS apps are native binaries (not JIT binaries)
+- CoreCLR runtime apps have no mechanism to publish universal binaries.
+
+### Dotnet-SOS and the SOS plugin Issues
+
+The `dotnet-sos` tool installs the `SOS` diagnostic plugin into `lldb`. Since `lldb` is a universal binary, it expects its plugins to also be universal binaries. With the single architecture design, when `SOS` is installed, `lldb` will complain about not being able to load the plugin when the architecture mismatches.
+
+#### Universal Binary Experiment
+
+Making the Apple Silcon SOS be a universal binary to allowed SOS to work on `x86_64` and `arm64` processes.
+
+## User Experience Designs
+
+### Mixed Architecture Side By Side Installs
```

Comment:
How are other runtimes/SDKs with global installs planning to address this problem?

It may be useful to look at Java, golang, python, rust, ... to see whether there is a trend.

Reply:
- I couldn't find an Oracle JDK for Apple Silicon.
- Zulu OpenJDK was arch specific.
- golang was arch specific.
- Python a universal binary. I think Apple provided the one I looked at
- Rust looks like it is not ready. They are still protoyping.

Reply:
Our muxer architecture, readytorun story, and legacy mixed architecture make it more complicated.

Reply:
From this list, it sounds like arch specific is the norm and we won't be an outlier if we stick with the simpler arch-specific as well.